<?xml version="1.0"?>

<launch>

    <arg name="localization_type" default="gps"/>

    <!-- Start Webots -->
    <arg name="no_gui" default="false" doc="Start Webots with minimal GUI" />
    <include file="$(find bupt_webots)/launch/webots.launch">
        <arg name="mode" value="realtime" />
        <arg name="no_gui" value="$(arg no_gui)" />
        <arg name="world" value="$(find bupt_webots)/worlds/robocon2023.wbt" />
    </include>

    <!-- <group ns="omni_chassis">
        <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" respawn="true" respawn_delay="3" launch-prefix="bash -c 'sleep 5; $0 $@' ">
        </node>
        <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher" respawn="true" respawn_delay="3" launch-prefix="bash -c 'sleep 5; $0 $@' ">
            <param name="rate" value="60"/>
        </node>
    </group> -->

    <!-- <group ns="mecanum_chassis">
        <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" respawn="true" respawn_delay="3" launch-prefix="bash -c 'sleep 5; $0 $@' ">
        </node>
        <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher" respawn="true" respawn_delay="3" launch-prefix="bash -c 'sleep 5; $0 $@' ">
            <param name="rate" value="60"/>
        </node>
    </group> -->

    <rosparam file="$(find bupt_webots)/config/ros_control.yaml" command="load"/>
    <group ns="r1">
        <param name="robot_description" textfile="$(find bupt_webots)/urdf/robocon2022_r1.urdf"/>
        <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" respawn="true" respawn_delay="3" launch-prefix="bash -c 'sleep 5; $0 $@' ">
            <param name="tf_prefix" value="r1"/>
        </node>

        <!-- localization -->
        <node pkg="bupt_webots" type="localization_webots" name="localization_webots" output="screen">
            <param name="tf_prefix" value="r1"/>
            <param name="localization_type" value="$(arg localization_type)"/>
            <param name="publish_rate" value="60"/>
            <param name="publish_tf" value="true"/>
        </node>
        <!-- <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher" respawn="true" respawn_delay="3" launch-prefix="bash -c 'sleep 5; $0 $@' ">
            <param name="rate" value="60"/>
            <param name="tf_prefix" value="r1"/>
        </node> -->

        <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="true" output="screen" args="
            wheel_drive_velocity_controller
            wheel_steer_position_controller
            joint_state_controller
            " />
    </group>

    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find bupt_webots)/rviz/robocon2023.rviz" />
</launch>
